osu.Server.Queues.ScoreStatisticsProcessor/Commands/Performance/Scores/UpdateAllScoresCommand.cs

```diff
@@ -0,0 +1,120 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dapper;
+using McMaster.Extensions.CommandLineUtils;
+using osu.Server.QueueProcessor;
+using osu.Server.Queues.ScoreStatisticsProcessor.Helpers;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Commands.Performance.Scores
+{
+    [Command(Name = "all", Description = "Computes pp of all scores from all users.")]
+    public class UpdateAllScoresCommand : PerformanceCommand
+    {
+        private const int max_scores_per_query = 5000;
+
+        [Option(Description = "Score ID to start processing from.")]
+        public ulong From { get; set; }
+
+        /// <summary>
+        /// Whether to adjust processing rate based on slave latency. Defaults to <c>false</c>.
+        /// </summary>
+        [Option(CommandOptionType.SingleOrNoValue, Template = "--check-slave-latency")]
+        public bool CheckSlaveLatency { get; set; }
+
+        protected override async Task<int> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            using var db = DatabaseAccess.GetConnection();
+
+            ulong currentScoreId = From;
+            ulong? lastScoreId = await db.QuerySingleAsync<ulong>("SELECT MAX(id) FROM scores");
+
+            ulong processedCount = 0;
+            ulong changedPp = 0;
+            double rate = 0;
+            Stopwatch sw = new Stopwatch();
+
+            var scoresQuery = db.Query<SoloScore>("SELECT * FROM scores WHERE `id` > @ScoreId AND `id` <= @LastScoreId ORDER BY `id`", new
+            {
+                ScoreId = currentScoreId,
+                LastScoreId = lastScoreId,
+            }, buffered: false);
+
+            using var scoresEnum = scoresQuery.GetEnumerator();
+
+            Console.WriteLine($"Processing all scores up to {lastScoreId}, starting from {currentScoreId}");
+
+            Task<List<SoloScore>> nextScores = getNextScores();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (CheckSlaveLatency)
+                    await SlaveLatencyChecker.CheckSlaveLatency(db, cancellationToken);
+
+                sw.Restart();
+
+                var scores = await nextScores;
+                nextScores = getNextScores();
+
+                if (scores.Count == 0)
+                    break;
+
+                await Task.WhenAll(Partitioner.Create(scores).GetPartitions(Threads).Select(async partition =>
+                {
+                    using (var connection = DatabaseAccess.GetConnection())
+                    using (var transaction = await connection.BeginTransactionAsync(IsolationLevel.ReadUncommitted, cancellationToken))
+                    using (partition)
+                    {
+                        while (partition.MoveNext())
+                        {
+                            if (cancellationToken.IsCancellationRequested)
+                                return;

```

Comment:
Note that if this is cancelled, then you're going to be printing a `currentScoreId` that may not have actually been processed yet. `ProcessPartitioned` has the same issue.

You may want to check the cancellation after the `await Task.WhenAll()` and not continue printing, lest there be confusion and scores missed because it was restarted from the wrong point.

Reply:
https://github.com/ppy/osu-queue-score-statistics/pull/284/commits/92700c1d3ada0c3ee69951d80d74d869ffccb254